File: apina-core/src/main/java/fi/evident/apina/java/model/JavaClass.java
package fi.evident.apina.java.model;

import fi.evident.apina.java.model.type.JavaBasicType;
import fi.evident.apina.java.model.type.JavaType;
import fi.evident.apina.java.model.type.TypeSchema;
import org.objectweb.asm.Opcodes;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

import static java.util.Collections.unmodifiableList;
import static java.util.Objects.requireNonNull;
import static java.util.stream.Collectors.toList;

/**
 * Contains all information read about class.
 */
public final class JavaClass implements JavaAnnotatedElement {

    private final JavaType type;
    private final JavaType superClass;
    private final List<JavaType> interfaces;
    private final List<JavaAnnotation> annotations = new ArrayList<>();
    private final List<JavaField> fields = new ArrayList<>();
    private final List<JavaMethod> methods = new ArrayList<>();
    private final int modifiers;
    private final TypeSchema schema;

    public JavaClass(JavaType type, JavaType superClass, List<JavaType> interfaces, int modifiers, TypeSchema schema) {
        this.modifiers = modifiers;
        this.type = requireNonNull(type);
        this.superClass = requireNonNull(superClass);
        this.interfaces = unmodifiableList(requireNonNull(interfaces));
        this.schema = requireNonNull(schema);
    }

    public boolean isEnum() {
        return (modifiers & Opcodes.ACC_ENUM) != 0;
    }

    public String getName() {
        return type.getNonGenericClassName();
    }

    public JavaType getType() {
        return type;
    }

    public JavaType getSuperClass() {
        return superClass;
    }

    public List<JavaType> getInterfaces() {
        return unmodifiableList(interfaces);
    }

    public List<JavaField> getFields() {
        return unmodifiableList(fields);
    }

    public Stream<JavaField> getPublicFields() {
        return fields.stream().filter(JavaField::isPublic);
    }

    public List<JavaField> getPublicInstanceFields() {
        return getPublicFields().filter(f -> !f.isStatic()).collect(toList());
    }

    public List<JavaMethod> getGetters() {
        return getPublicMethods().filter(JavaMethod::isGetter).collect(toList());
    }

    public List<JavaMethod> getMethods() {
        return unmodifiableList(methods);
    }

    public Stream<JavaMethod> getPublicMethods() {
        return methods.stream().filter(JavaMethod::isPublic);
    }

    @Override
    public List<JavaAnnotation> getAnnotations() {
        return unmodifiableList(annotations);
    }

    public void addAnnotation(JavaAnnotation annotation) {
        annotations.add(requireNonNull(annotation));
    }

    public void addField(JavaField field) {
        fields.add(requireNonNull(field));
    }

    public void addMethod(JavaMethod method) {
        methods.add(requireNonNull(method));
    }

    /**
     * Returns enum constants defined by this class, if this class is an enum.
     *
     * @return Names of enum constants, in the order they appear
     * @throws UnsupportedOperationException if the class is not an enum
     */
    public List<String> getEnumConstants() {
        if (!isEnum()) throw new UnsupportedOperationException("not an enum: " + this);

        List<String> result = new ArrayList<>(fields.size());
        for (JavaField field : fields)
            if (field.isEnumConstant())
                result.add(field.getName());
        return result;
    }

    @Override
    public String toString() {
        return type.toString();
    }

    public JavaField getField(String name) {
        JavaField field = fields.stream().filter(f -> name.equals(f.getName())).findFirst().orElse(null);
        if (field != null)
            return field;
        else
            throw new RuntimeException("field not found " + name);
    }

    public TypeSchema getSchema() {
        return schema;
    }

    public boolean hasMethodWithAnnotation(JavaBasicType annotationType) {
        return methods.stream().anyMatch(m -> m.hasAnnotation(annotationType));
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/java/model/JavaClass.kt
package fi.evident.apina.java.model

import fi.evident.apina.java.model.type.JavaType
import fi.evident.apina.java.model.type.TypeSchema
import kotlinx.metadata.KmClass
import kotlinx.metadata.jvm.KotlinClassMetadata
import kotlinx.metadata.jvm.Metadata
import org.objectweb.asm.Opcodes
import java.lang.reflect.Modifier

/**
 * Contains all information read about class.
 */
class JavaClass(
    val type: JavaType,
    val superClass: JavaType,
    val interfaces: List<JavaType>,
    private val modifiers: Int,
    val schema: TypeSchema
) : JavaAnnotatedElement {

    private val _annotations = ArrayList<JavaAnnotation>()
    private val _fields = ArrayList<JavaField>()
    private val _methods = ArrayList<JavaMethod>()
    private val _recordComponent = ArrayList<JavaRecordComponent>()

    override val annotations: List<JavaAnnotation>
        get() = _annotations

    val methods: List<JavaMethod>
        get() = _methods

    val fields: List<JavaField>
        get() = _fields

    val recordComponents: List<JavaRecordComponent>
        get() = _recordComponent

    val isEnum: Boolean
        get() = modifiers and Opcodes.ACC_ENUM != 0

    val isRecord: Boolean
        get() = modifiers and Opcodes.ACC_RECORD != 0

    val isAnnotation: Boolean
        get() = isInterface && interfaces.contains(ANNOTATION_TYPE)

    val isInterface: Boolean
        get() = Modifier.isInterface(modifiers)

    val name: String
        get() = type.nonGenericClassName

    private val publicFields: Sequence<JavaField>
        get() = fields.asSequence().filter { it.isPublic }

    val publicInstanceFields: List<JavaField>
        get() = publicFields.filter { !it.isStatic }.toList()

    val getters: List<JavaMethod>
        get() = publicMethods.filter { it.isGetter }.toList()

    val publicMethods: Sequence<JavaMethod>
        get() = methods.asSequence().filter { it.isPublic }

    fun addAnnotation(annotation: JavaAnnotation) {
        _annotations += annotation
    }

    fun addField(field: JavaField) {
        _fields += field
    }

    fun addMethod(method: JavaMethod) {
        _methods += method
    }

    fun addRecordComponent(component: JavaRecordComponent) {
        _recordComponent += component
    }

    /**
     * Returns enum constants defined by this class, if this class is an enum.
     *
     * @return Names of enum constants, in the order they appear
     * @throws IllegalStateException if the class is not an enum
     */
    val enumConstants: List<String>
        get() {
            check(isEnum) { "not an enum: $this" }

            return fields.filter { it.isEnumConstant }.map { it.name }
        }

    val kotlinMetadata: KmClass? by lazy {
        findAnnotation(KOTLIN_METADATA_ANNOTATION)?.let { metadata ->
            val header = Metadata(
                kind = metadata.getAttribute("k"),
                metadataVersion = metadata.getAttribute("mv"),
                data1 = metadata.getAttribute("d1"),
                data2 = metadata.getAttribute("d2"),
                extraString = metadata.getAttribute("xs"),
                packageName = metadata.getAttribute("pn"),
                extraInt = metadata.getAttribute("xi")
            )
            (KotlinClassMetadata.readLenient(header) as? KotlinClassMetadata.Class)?.kmClass
        }
    }

    override fun toString() = type.toString()

    fun getField(name: String): JavaField =
        findField(name) ?: throw RuntimeException("field not found $name")

    fun findField(name: String): JavaField? =
        fields.find { name == it.name }

    fun findMethodWithAnnotation(annotationType: JavaType.Basic) =
        methods.find { it.hasAnnotation(annotationType) }

    companion object {
        private val ANNOTATION_TYPE = JavaType.Basic(Annotation::class.java)
        private val KOTLIN_METADATA_ANNOTATION = JavaType.Basic(Metadata::class.java)
    }
}
